libvast/builtins/formats/json.cpp

```diff
@@ -0,0 +1,356 @@
+        = parse_impl(element.value_unsafe(), list, depth + 1))
+      return parsed_element;
+  }
+  return {};
+}
+
+parse_result
+parse_object(simdjson::ondemand::value val, auto& pusher, size_t depth) {
+  auto result = val.get_object();
+  if (result.error())
+    return result.error();
+  auto obj = result.value_unsafe();
+  auto record = pusher.push_record();
+  for (auto pair : obj) {
+    if (pair.error())
+      return pair.error();
+    auto key = pair.unescaped_key();
+    if (key.error())
+      return key.error();
+    auto val = pair.value();
+    if (val.error())
+      return val.error();
+    auto field = record.push_field(key.value_unsafe());
+    if (auto parse_result = parse_impl(val.value_unsafe(), field, depth + 1))
+      return parse_result;
+  }
+  return {};
+}
+
+parse_result
+parse_impl(simdjson::ondemand::value val, auto& pusher, size_t depth) {
+  if (depth > defaults::max_recursion)
+    die("nesting too deep in json_parser parse");
+  auto type = val.type();
+  if (type.error())
+    return type.error();
+  switch (val.type().value_unsafe()) {
+    case simdjson::ondemand::json_type::null:
+      return {};
+    case simdjson::ondemand::json_type::number:
+      return parse_number(val, pusher);
+    case simdjson::ondemand::json_type::boolean: {
+      auto result = val.get_bool();
+      if (result.error())
+        return result.error();
+      pusher.add(result.value_unsafe());
+      return {};
+    }
+    case simdjson::ondemand::json_type::string:
+      return parse_string(val, pusher);
+    case simdjson::ondemand::json_type::array:
+      return parse_array(val, pusher, depth + 1);
+    case simdjson::ondemand::json_type::object:
+      return parse_object(val, pusher, depth + 1);
+  }
+  die("unhandled json object type in switch statement");
+}
+
+parse_result parse_entry(simdjson::ondemand::object object,
+                         adaptive_table_slice_builder::row_guard& row) {
+  for (auto pair : object) {
+    if (pair.error())
+      return pair.error();
+    auto key = pair.unescaped_key();
+    if (key.error())
+      return key.error();
+    auto val = pair.value();
+    if (val.error())
+      return val.error();
+    auto field = row.push_field(key.value_unsafe());
+    if (auto parse_result = parse_impl(val.value_unsafe(), field, 0u))
+      return parse_result;
+  }
+  return {};
+}
+
+parse_result
+parse(simdjson::simdjson_result<simdjson::ondemand::document_reference> doc,
+      adaptive_table_slice_builder& builder) {
+  if (doc.error())
+    return doc.error();
+  auto obj = doc.value_unsafe().get_object();
+  if (obj.error())
+    return obj.error();
+  auto row = builder.push_row();
+  if (auto err = parse_entry(obj.value_unsafe(), row)) {
+    row.cancel();
+    return err;
+  }
+  return {};
+}
+
+void emit_warning(operator_control_plane& control_plane,
+                  simdjson::error_code err) {
+  control_plane.warn(caf::make_error(ec::parse_error, error_message(err)));
+}
```

Comment:
This makes it so we always just forward the simdjson warnings directly to the user, which are not written in a user-facing way. Can you either change this, or write a follow-up issue to improve the error messages?

Reply:
I will write the followup

Reply:
Followup story created: https://github.com/tenzir/issues/issues/274